openpype/hosts/fusion/api/plugin.py

```diff
@@ -0,0 +1,257 @@
+        workdir = os.path.normpath(legacy_io.Session["AVALON_WORKDIR"])
+        formatting_data.update({
+            "workdir": workdir,
+            "frame": "0" * frame_padding,
+            "ext": ext,
+            "product": {
+                "name": formatting_data["subset"],
+                "type": formatting_data["family"],
+            },
+        })
+
+        # build file path to render
+        filepath = self.temp_rendering_path_template.format(**formatting_data)
+
+        comp = get_current_comp()
+        tool["Clip"] = comp.ReverseMapPath(os.path.normpath(filepath))
+
+        # Rename tool
+        if tool.Name != subset:
+            print(f"Renaming {tool.Name} -> {subset}")
+            tool.SetAttrs({"TOOLS_Name": subset})
+
+    def get_managed_tool_data(self, tool):
+        """Return data of the tool if it matches creator identifier"""
+        data = tool.GetData("openpype")
+        if not isinstance(data, dict):
+            return
+
+        required = {
+            "id": "pyblish.avalon.instance",
+            "creator_identifier": self.identifier,
+        }
+        for key, value in required.items():
+            if key not in data or data[key] != value:
+                return
+
+        # Get active state from the actual tool state
+        attrs = tool.GetAttrs()
+        passthrough = attrs["TOOLB_PassThrough"]
+        data["active"] = not passthrough
+
+        # Override publisher's UUID generation because tool names are
+        # already unique in Fusion in a comp
+        data["instance_id"] = tool.Name
+
+        return data
+
+    def get_pre_create_attr_defs(self):
+        """Settings for create page"""
+        raise NotImplementedError
+
+    def get_instance_attr_defs(self):
+        """Settings for publish page"""
+        return self.get_pre_create_attr_defs()
+
+    def get_default_variants(self):
+        """Calculate variants from Settings and loaded containers.
+
+        Tries to parse out variant from loaded containers.
+        """
+        variants = super().get_default_variants()
+
+        for container in self.host.get_containers():
+            repre_id = container.get("representation")
+            if not repre_id:
+                continue
+            repre_doc = get_representation_by_id(self.project_name,
+                                                 repre_id)
+            if not repre_doc:
+                self.log.warning(f"No representation for {repre_id}")
+                continue
+
+            context = repre_doc["data"]["context"]
+            product_info = context["product"]
+            variant = product_info["name"]
+            # try to get "clean" variant from loaded product name
+            for replace_str in [product_info["name"], context["asset"],
+                                context["task"]["name"]]:
+                variant = re.sub(replace_str, "",
+                                 variant, flags=re.I)
+            variants.append(variant)
```

Comment:
This will **always** run on reset of the publisher - and this query may be slow - are we sure this is what we want to do? This will slow down the reset of the publisher quite a bit.

At the very least we should avoid the `get_representation_by_id` query per loaded container but batch them together. We easily have 50+ containers in our comps - even on small-to-mid-size projects.

Also, the `re.sub` is likely much faster if not initializing the `re.sub` per replace string but instead make them an _or_ statement so that it matches any of them at once in the state machine.

```python
pattern = "|".join(words)
re.sub(pattern, "", variant, flags=re.IGNORECASE)
```

Also note that it's easier to read when using the long flag `re.IGNORECASE` than `re.I` so I'd recommend that too.

---

Anyway, if we want to do matching on `variants` this shows that that is _actually_ data we should be storing on the published version. This code here is hacky and unreliable - it does not seem to adhere to defined templates at all (and the templates might even not use `variant`!) - as such. This workflow and code seems either:

- Client specific and work only under (very) specific conditions
- Or is built wrong and we should instead start producing the actual data in the published content so we can build this feature more reliably.